### ISSUE TITLE
Align the SDK with the new DSL

### ISSUE THREAD

**Issue by spolti** [serverlessworkflow/sdk-go#203]
**What would you like to be added**:

The DSL has been updated to a new and concise version.

**Why is this needed**:
The SDK needs to support the new DSL, as found here: https://github.com/serverlessworkflow/specification/blob/main/dsl-reference.md

For more info: https://github.com/serverlessworkflow/specification/issues/843

**Comment by ribeiromiranda** [comment]
What is your opinion on generating the SDK code based on the json schema?https://github.com/serverlessworkflow/specification/blob/main/schema/workflow.yaml

**Comment by ricardozanini** [comment]
@ribeiromiranda +1 if we can, although in my experience, the workflow can be really complex and hard to model using generators. We have much `OneOf` and `AnyOf` that makes everything harder to model in Go.

**Comment by ribeiromiranda** [comment]
I understand, generating code can get very complex.

Another very interesting way to make a json schema interpreter with libraries: https://cuelang.org or https://github.com/xeipuuv/gojsonschema.

Is it simpler than generating the code and robust to changes in the specification.

**Comment by ricardozanini** [comment]
+1, are you willing to give it a try?

**Comment by fjtirado** [comment]
The DSL is still changing, so I would wait till is more stable.

**Comment by ricardozanini** [comment]
@fjtirado despite the stability of the DSL, it's worth starting exploring our possibilities first.

**Comment by fjtirado** [comment]
> oints, unfortunately, Go doesn't play well with inheritance. Let's see what we can do given @ribeiromiranda suggestions. I agree that in the Java SDK option 3 can be a good alternative for the `xOf` approach.

Yes, for option 3. in go, the option is to wrap the generated Pojo into another one https://www.geeksforgeeks.org/inheritance-in-golang/ (since GO allows calling the methods of the generated pojo within the container one, it will do the trick)

**Comment by ribeiromiranda** [comment]
I think these features below can be using jsonschema validation:
* Parse workflow JSON and YAML definitions
* Validate workflow definitions (Schema)
  * Custom error messages
<br>

Development the validation (e.g. map jsonschema into a graph):
* Validate workflow definitions (Integrity)
* Generate workflow diagram (SVG)
<br>

Generate code based in jsonschema
* Programmatically build workflow definitions
<br>
<br>


I don't know if it has other features or I might be simplistic? I think there is no need to map jsonschema into struct. If everyone agrees, I can try to develop some simple cases.

 A very simple validation example:
```go
package main

import (
	_ "embed"
	"fmt"
	"log"
	"os"

	"github.com/xeipuuv/gojsonschema"
	"sigs.k8s.io/yaml"
)

//go:embed workflow.yaml
var yamlWorkflow []byte

func main() {

	jsonWorkflow, err := yaml.YAMLToJSON(yamlWorkflow)
	if err != nil {
		log.Fatal(err.Error())
	}

	yamlExample, err := os.ReadFile("./example.yaml")
	if err != nil {
		log.Fatal(err.Error())
	}
	jsonExample, err := yaml.YAMLToJSON(yamlExample)
	if err != nil {
		log.Fatal(err.Error())
	}

	schemaLoader := gojsonschema.NewBytesLoader(jsonWorkflow)
	documentLoader := gojsonschema.NewBytesLoader(jsonExample)

	result, err := gojsonschema.Validate(schemaLoader, documentLoader)
	if err != nil {
		log.Fatal(err.Error())
	}

	if result.Valid() {
		fmt.Printf("The document is valid\n")
	} else {
		fmt.Printf("The document is not valid. see errors :\n")
		for _, desc := range result.Errors() {
			fmt.Printf("- %s\n", desc)
		}
	}
}
```

**Comment by ricardozanini** [comment]
> > oints, unfortunately, Go doesn't play well with inheritance. Let's see what we can do given @ribeiromiranda suggestions. I agree that in the Java SDK option 3 can be a good alternative for the `xOf` approach.
> 
> Yes, for option 3. in go, the option is to wrap the generated Pojo into another one https://www.geeksforgeeks.org/inheritance-in-golang/ (since GO allows calling the methods of the generated pojo within the container one, it will do the trick)

I know how to do inheritance in Go, but that's not the point. The point is this technique is an anti-pattern in general we avoid.

**Comment by fjtirado** [comment]
@ribeiromiranda I think that the input to the validator should be just the workflow definition stream (an io.reader) being validated (the loading of the validating schema should be hidden to the developer)

**Comment by spolti** [comment]
just a reminder about it:

> Just provide validation that certain definition file is valid according to the schema (using https://github.com/xeipuuv/gojsonschema) . And maybe a shortcut call to load the generic Json object (load the map[string]interface{} as in [here](https://go.dev/blog/json)) representing the workflow on memory

using generics with the SDK is a not good approach due the kubernetes integration. we had some problem related to it in the past.

**Comment by ricardozanini** [comment]
I don't see an issue with this representation, @ribeiromiranda @spolti. And I agree that generics is a no-go. Too much compatibility issues with k8s libraries.

**Comment by ribeiromiranda** [comment]
I created the [repository](https://github.com/galgotech/sdk-go-1.0) with a proposal to implement the new DSL, if I follow this implementation I will create a PR.

This repository implemented a basic "validator" using jsonschema and "builder", in the file "example/example.go" there are some use cases.

https://github.com/galgotech/sdk-go-1.0

**Comment by ricardozanini** [comment]
@ribeiromiranda sorry for the late reply, I was on PTO. 
I liked the approach, but I only scanned the repo. I think it would be useful to send the PR so we can discuss/have a throughout review. 

To me it seems aligned to what we talked.

**Comment by ricardozanini** [comment]
One little thing: this is supposed to be v4, not v3 since we are skipping 0.9 release.

**Comment by paulmchen** [comment]
@ricardozanini Ricardo, I have a quick question about the current support for Go SDK. What is the latest version of the spec that the Go SDK supports. Thanks a lot.

**Comment by ricardozanini** [comment]
@paulmchen the latest one is 0.8, as highlighted in our README: https://github.com/serverlessworkflow/sdk-go?tab=readme-ov-file#releases

I'm on the verge of sending a PR for the next major version (v3) supporting 1.0.0

**Comment by paulmchen** [comment]
That's great! I'm looking forward to picking up the update! Thanks @ricardozanini

**Comment by ricardozanini** [comment]
@paulmchen Bear in mind that's the first iteration. I'll test it against our major examples directories, but bugs might be around.

**Comment by ricardozanini** [comment]
@paulmchen are you using the 0.8 one?

**Comment by paulmchen** [comment]
We are building our cloud workflow engine and would like to implement using the latest spec of the SWS.

**Comment by ricardozanini** [comment]
@paulmchen, when you have the opportunity, would you mind sharing the use case/usage with the community? We are about to apply for incubation, and it would be nice to hear your story.
